### ISSUE TITLE
Not authenticating on Google App Script

### ISSUE THREAD

**Issue by witeshadow** [toggl/toggl_api_docs#128]
I am trying to get this Google App script to work. It seems to have broken when the API was changed. 

What I added and am trying to get working is:  *function getworkspaces2()*

```
/*
    Imports time from www.toggl.com to a spreadsheet in Google Drive
    Copyright (C) 2013  koen-github

    This program is free software: you can redistribute it and/or modify
    it under the terms of the GNU General Public License as published by
    the Free Software Foundation, either version 3 of the License, or
    (at your option) any later version.

    This program is distributed in the hope that it will be useful,
    but WITHOUT ANY WARRANTY; without even the implied warranty of
    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
    GNU General Public License for more details.

    You should have received a copy of the GNU General Public License
    along with this program.  If not, see <http://www.gnu.org/licenses/>.

    Source located here: https://github.com/koen-github/Toggl-script-for-Google-Docs

*/
function onOpen() {
  var ss = SpreadsheetApp.getActiveSpreadsheet();
  var menuEntries = [ {name: "Get Times", functionName: "action"}                  ];
  ss.addMenu("Toggl Times", menuEntries);
}

var mydoc = SpreadsheetApp.getActiveSpreadsheet();
var sheet = mydoc.getSheets()[0];
var myCell = mydoc.getActiveCell();
var myRow = myCell.getRow();
var myCol = myCell.getColumn();

function action() {

    ScriptProperties.setProperties({
        'col': myCol,
        'row': myRow
    }, true);


    var app = UiApp.createApplication().setTitle('Timing from Last week - Select Time').setHeight(400).setWidth(500);

    var form = app.createFormPanel().setId('form').setEncoding('multipart/form-data');
    var formContent = app.createGrid().resize(15, 3);
    form.add(formContent);

    var unamepass = "xxxxxxxxxxxxxxxxxxxxxx"; //insert your api key here
    var digest = Utilities.base64Encode(unamepass);
    var digestfull = "Basic " + digest;


    url = "https://toggl.com/reports/api/v2time_entries.json";
   // options = 
    var response = UrlFetchApp.fetch(url, {
 // var response = UrlFetchApp.fetch("https://www.toggl.com/api/v6/time_entries.json", {

        method: "get",
        headers: {
            "Authorization": digestfull
        }
    });
    var test = response.getContentText();
    var blaat = Utilities.jsonParse(test).data;


    for (var i = 1; i <= blaat.length; i++) {
        formContent.setWidget(i, 0, app.createLabel(blaat[i - 1]["description"]));
        formContent.setWidget(i, 1, app.createRadioButton("radio1", cheat(blaat[i - 1]["duration"])).setFormValue(cheat(blaat[i - 1]["duration"])));
    }
    formContent.setWidget(13, 0, app.createSubmitButton('Insert Time!'));


    app.add(form);
    mydoc.show(app);


}

function doPost(e) {
    var row1 = ScriptProperties.getProperty('row');

    var col1 = ScriptProperties.getProperty('col');

    var app = UiApp.getActiveApplication();

    var radio = e.parameter.radio1;

    sheet.getRange(row1, col1, 1, 1).setValue(radio);
  return app.close()

}

function cheat(secs) {
    var t = new Date(1970, 0, 1);
    t.setSeconds(secs);
    return t.toTimeString().substr(0, 8);
}



function getworkspaces2() {
    var app = UiApp.createApplication().setTitle('Timing from Last week - Select Time').setHeight(400).setWidth(500);
    ScriptProperties.setProperties({
        'col': myCol,
        'row': myRow
    }, true);

    var form = app.createFormPanel().setId('form').setEncoding('multipart/form-data');
    var formContent = app.createGrid().resize(15, 3);
    form.add(formContent);
    
    var api_token = 'xxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxx';
    var unamepass = 'xxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxx:API_TOKEN'; //insert your api key here
    var digest = Utilities.base64Encode(unamepass);
    var digestfull = "Basic " + digest;


    var url_path = 'https://www.toggl.com/api/v8/workspaces';
   // options = 
   //var response = UrlFetchApp.fetch(dump);
       
       
       var extra = {
//     var extra = {headers: {
   //  var extra = {'payload': {
 // var response = UrlFetchApp.fetch("https://www.toggl.com/api/v6/time_entries.json", {
       method: 'GET',
//        headers: {
      'Content-Type': "application/json",
//        "Authorization": unamepass
        'Authorization': digestfull,
        user_agent: 'dummy@emails.com'
//          "user": api_token,
//          'password': 'API_TOKEN'
//        }
    }
    var result = UrlFetchApp.getRequest(url_path, extra);
//    Browser.msgBox(result);
//    Logger.log(result);
    var response = UrlFetchApp.fetch(url_path, extra);
//    Logger.log(response);
    var test = response.getContentText();
//    Logger.log(test);
    var blaat = Utilities.jsonParse(test).data;
//    Logger.log(blaat);
//    Browser.msgBox(blaat);
}
```

**Comment by refiito** [comment]
What you're probably interested in is 
https://github.com/toggl/toggl_api_docs/blob/master/reports/detailed.md
This endpoint gives you tons of information you'd probably want/like to display in the calendar

**Comment by witeshadow** [comment]
I'm not to that point yet. I get  404 when I run just to list projects.

**Comment by refiito** [comment]
Check the URL you're querying (the example code does not contain any project fetching).
The code, you have in the opening comment, is rather unreadable... 

Dumb it down and copy only the relevant part you're having trouble with.. also are the queries OK when you run them via curl ?

**Comment by witeshadow** [comment]
OK. this is much better. 

```
function getworkspaces2() {
    var app = UiApp.createApplication().setTitle('Timing from Last week - Select Time').setHeight(400).setWidth(500);
    ScriptProperties.setProperties({
        'col': myCol,
        'row': myRow
    }, true);

    var form = app.createFormPanel().setId('form').setEncoding('multipart/form-data');
    var formContent = app.createGrid().resize(15, 3);
    form.add(formContent);

    var api_token = 'xxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxx';
    var unamepass = 'xxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxx:API_TOKEN';
    var digest = Utilities.base64Encode(unamepass);
    var digestfull = "Basic " + digest;

    var url_path = 'https://www.toggl.com/api/v8/workspaces';
        var extra = {
       UrlFetchApp.fetch("https://www.toggl.com/api/v6/time_entries.json", {
       method: 'GET',
      'Content-Type': "application/json",
        'Authorization': digestfull,
        user_agent: 'dummy@emails.com'
    }
    var result = UrlFetchApp.getRequest(url_path, extra);
    var response = UrlFetchApp.fetch(url_path, extra);
    var test = response.getContentText();
    var blaat = Utilities.jsonParse(test).data;
}
```

**Comment by refiito** [comment]
The code you have does not fetch projects...
Also, only thing it seems to do is try to fetch something from https://www.toggl.com/api/v6/time_entries.json

API v6 has been deprecated for a few years now (and completely dropped in the end of last year), so, yes, 404 will be yielded there.

You should port the code to use the detailed report endpoint from reports api (https://github.com/toggl/toggl_api_docs/blob/master/reports/detailed.md)

Also, running queries with curl helps a lot in debugging....
